File: lib/pacer/pipe/block_filter_pipe.rb
module Pacer::Pipes
  class BlockFilterPipe < AbstractPipe
    field_reader :starts

    attr_reader :block

    def initialize(back, block, invert = false)
      super()
      @block = Pacer::Wrappers::WrappingPipeFunction.new back, block
    end

    def processNextStart()
      while raw_element = starts.next
        ok = block.call raw_element
        ok = !ok if @invert
        return raw_element if ok
      end
      raise EmptyPipe.instance
    end
  end
end

File: ext/src/main/java/pacer/pipe/BlockFilterPipe.java
package com.xnlogic.pacer.pipes;

import com.tinkerpop.pipes.AbstractPipe;


public class BlockFilterPipe<T> extends AbstractPipe<T, T> {
    private boolean invert;

    // TODO: Figure out if this is using Java 7 or 8 as Java 7 doesn't support lambdas/blocks.
    /*
    public BlockFilterPipe(back, block, boolean invert) {
        super();

        this.invert = invert;
        this.block = ;
    }

    public BlockFilterPipe(back, block) {
        this(back, block, false);
    }
*/
    protected T processNextStart() {
        return this.starts.next();    
    }
}
